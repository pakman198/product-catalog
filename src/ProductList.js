import React from 'react';

import CatalogContext from './CatalogContext';
import Product from './Product';

const ProductList = () => {

  const renderProducts = (products) => {
    return products.map(product => {
      return (
        <div key={product.id} className="four wide column">
          <Product data={product} />
        </div>
      );
    });
  }

  return (
    <CatalogContext.Consumer>
      {
        context => (
          <div className="ui grid">
            { renderProducts(context.products) }
          </div>
        )
      }
    </CatalogContext.Consumer>
  );
}

export default ProductList;