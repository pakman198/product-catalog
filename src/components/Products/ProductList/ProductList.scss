.ui.fixed.menu+.ui.grid {
  &.product-list {
    padding-top: 4.75rem;
  }
}