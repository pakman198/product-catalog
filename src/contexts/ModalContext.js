import React, { useState } from 'react';

export const ModalContext = React.createContext();

export const ModalProvider = (props) => {
  const [ isModalVisible, setIsModalVisible ] = useState(false);
  const [ currentItem, setCurrentItem ] = useState({});
  

  const toggleModal = (item) => {
    const selectedItem = item ? item : null;

    setIsModalVisible(!isModalVisible);
    setCurrentItem(selectedItem);
  }

  const contextValue = {
    currentItem,
    isModalVisible,
    toggleModal
  }
    
  return (
    <ModalContext.Provider value={contextValue}>
      { props.children }
    </ModalContext.Provider>
  );
}

