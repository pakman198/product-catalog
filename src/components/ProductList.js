import './ProductList.scss';
import React, { Component } from 'react';
import axios from 'axios';

import Product from './Product';
import Modal from './Modal';
import ModalContext from '../ModalContext';

class ProductList extends Component {
  static contextType = ModalContext;

  state = {
    products: []
  }

  componentDidMount() {
    axios.get('http://localhost:3001/').then(({data}) => {
      this.setState({ products: data.groups })
    }).catch(e => {
      console.log(e);
    })
  }

  renderProducts() {
    const { products } = this.state;
    if ( !products || products.length === 0) return null;

    return products.map(product => {
      return (
        <div key={product.id} className="column">
          <Product data={product} />   
        </div>
      )
    })
  }

  renderModal() {
    const { toggleModal, currentItem: { name, images }} = this.context;

    return (
      <Modal title={name} images={images} onDismiss={toggleModal} />
    )
  }

  render() {
    const products = this.renderProducts();
    // console.log(this.context)

    return (
      <>
        <div className="product-list ui stackable three column grid">
          { products }
        </div>
        { this.context.isModalVisible ? this.renderModal() : null }
      </>
    );
  }
}

export default ProductList;