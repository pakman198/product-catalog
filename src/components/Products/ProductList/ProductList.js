import './ProductList.scss';
import React, { useState, useContext, useEffect } from 'react';
import axios from 'axios';

import Product from '../Product';
import Modal from '../../Modal/Modal';
import { ModalContext } from '../../../contexts/ModalContext';

const ProductList = () => {
  const modalContext = useContext(ModalContext);
  const [ products, setProducts ] = useState([]);

  useEffect(() => {
    axios.get('http://localhost:3001/').then(({data}) => {
      setProducts(data.groups);
    }).catch(e => {
      console.log(e);
    })
  }, [])

  let items = null;
  if ( products || products.length !== 0) {
    items =  products.map(product => {
      return (
        <div key={product.id} className="column">
          <Product data={product} />   
        </div>
      );
    });
  }

  return (
    <>
      <div className="product-list ui stackable three column grid">
        { items }
      </div>
      { modalContext.isModalVisible ? <Modal /> : null }
    </>
  );
  
}

export default ProductList;