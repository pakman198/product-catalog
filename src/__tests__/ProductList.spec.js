import React from 'react';
import axiosMock from "axios";
import { 
  render, 
  waitForElement, 
} from '@testing-library/react'
import '@testing-library/jest-dom';

import ProductList from '../components/ProductList';
import ModalContext from '../ModalContext';
import data from '../__mocks__/apiResponse.json';

describe('ProductList', () => {

  const context = {
    currentItem: null,
    isModalVisible: false,
    toggleModal: jest.fn()
  }

  const context2 = {
    currentItem: data[0],
    isModalVisible: true,
    toggleModal: jest.fn()
  }


  test("Is displayed properly", () => {
    axiosMock.get.mockResolvedValueOnce({ data: { groups: data }});
    const { asFragment } = render(
      <ModalContext.Provider value ={context}>
        <ProductList />
      </ModalContext.Provider>
    );
    
    expect(asFragment()).toMatchSnapshot();
  });

  test('Displays a Modal if the value from the context is true', async () => {
    axiosMock.get.mockResolvedValueOnce({ data: { groups: data }});
    
    const modalRoot = document.createElement('div');
    modalRoot.setAttribute('id', 'modal');
    document.body.appendChild(modalRoot)
    
    const { getByText } = render(
      <ModalContext.Provider value ={context2}>
        <ProductList />
      </ModalContext.Provider>
    );

    const modal = await waitForElement(() => getByText(/Tahoe Solid Wood Dining Table - Natural Oak/i));

    expect(modal).toBeInTheDocument;
  });

});

