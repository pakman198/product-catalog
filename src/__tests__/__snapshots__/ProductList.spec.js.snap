// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList Is displayed properly 1`] = `
<DocumentFragment>
  <div
    class="product-list ui stackable three column grid"
  >
    <div
      class="column"
    >
      <div
        class="ui centered card"
      >
        <div
          class="image"
          role="button"
        >
          <img
            alt="Tahoe Solid Wood Dining Table - Natural Oak"
            src="https://www.westelm.com/weimgs/rk/images/wcm/products/201938/0025/tahoe-solid-wood-dining-table-natural-oak-m.jpg"
          />
        </div>
        <div
          class="content"
        >
          <div
            class="header"
          >
            Tahoe Solid Wood Dining Table - Natural Oak
          </div>
        </div>
        <div
          class="extra content"
        >
          $1,359.20
        </div>
      </div>
    </div>
    <div
      class="column"
    >
      <div
        class="ui centered card"
      >
        <div
          class="image"
          role="button"
        >
          <img
            alt="Frame Dining Table - Caramel/Antique Bronze"
            src="https://www.westelm.com/weimgs/rk/images/wcm/products/201938/0012/frame-dining-table-caramel-antique-bronze-m.jpg"
          />
        </div>
        <div
          class="content"
        >
          <div
            class="header"
          >
            Frame Dining Table - Caramel/Antique Bronze
          </div>
        </div>
        <div
          class="extra content"
        >
          $559.20
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
